rcldotnet/Clock.cs

```diff
@@ -0,0 +1,261 @@
+            _dllLoadUtils.RegisterNativeFunction(nativeLibrary, nameof(native_rcl_set_ros_time_override), out native_rcl_set_ros_time_override);
+            _dllLoadUtils.RegisterNativeFunction(nativeLibrary, nameof(native_rcl_clock_add_jump_callback), out native_rcl_clock_add_jump_callback);
+            _dllLoadUtils.RegisterNativeFunction(nativeLibrary, nameof(native_rcl_clock_remove_jump_callback), out native_rcl_clock_remove_jump_callback);
+        }
+    }
+
+    public sealed class Clock
+    {
+        private readonly Dictionary<JumpCallback, JumpCallbackInternal> _registeredJumpCallbacks = new Dictionary<JumpCallback, JumpCallbackInternal>();
+
+        internal Clock(SafeClockHandle handle)
+        {
+            Handle = handle;
+        }
+
+        internal SafeClockHandle Handle { get; }
+
+        public Time Now()
+        {
+            RCLRet ret = ClockDelegates.native_rcl_clock_get_now(Handle, out TimePoint timePoint);
+
+            RCLExceptionHelper.CheckReturnValue(ret, $"{nameof(ClockDelegates.native_rcl_clock_get_now)}() failed.");
+
+            return timePoint.ToMsg();
+        }
+
+        internal RCLRet EnableRosTimeOverride()
+        {
+            RCLRet ret = ClockDelegates.native_rcl_enable_ros_time_override(Handle);
+
+            RCLExceptionHelper.CheckReturnValue(ret, $"{nameof(ClockDelegates.native_rcl_enable_ros_time_override)}() failed.");
+
+            return ret;
+        }
+
+        internal RCLRet DisableRosTimeOverride()
+        {
+            RCLRet ret = ClockDelegates.native_rcl_disable_ros_time_override(Handle);
+
+            RCLExceptionHelper.CheckReturnValue(ret, $"{nameof(ClockDelegates.native_rcl_disable_ros_time_override)}() failed.");
+
+            return ret;
+        }
+
+        internal RCLRet SetRosTimeOverride(long timePointValue)
+        {
+            RCLRet ret = ClockDelegates.native_rcl_set_ros_time_override(Handle, timePointValue);
+
+            RCLExceptionHelper.CheckReturnValue(ret, $"{nameof(ClockDelegates.native_rcl_set_ros_time_override)}() failed.");
+
+            return ret;
+        }
+
+        public void AddJumpCallback(JumpThreshold threshold, JumpCallback callback)
+        {
+            if (_registeredJumpCallbacks.ContainsKey(callback))
+            {
+                throw new CallbackAlreadyRegisteredException("Provided jump callback was already registered!");
+            }
+
+            JumpCallbackInternal callbackInternal = (timeJumpPtr, beforeJump) =>
+                callback(Marshal.PtrToStructure<TimeJump>(timeJumpPtr), beforeJump);
+
+            RCLRet ret = ClockDelegates.native_rcl_clock_add_jump_callback(Handle, threshold, callbackInternal);
+
+            RCLExceptionHelper.CheckReturnValue(ret, $"{nameof(ClockDelegates.native_rcl_clock_add_jump_callback)}() failed.");
+
+            _registeredJumpCallbacks.Add(callback, callbackInternal);
+        }
+
+        public bool RemoveJumpCallback(JumpCallback callback)
```

Comment:
I've been looking into implementing this suggestion, but I'm having trouble reconciling it with the request to add an instanced lock for thread safety in a way which makes sense. In order to dispose of itself, this object would need a reference to both the `SafeClockHandle` and the `_lock` object or the add and remove functions would need to be wrapped in methods supplied by `Clock` in which case only a `Clock` reference is needed.

Here's a Gist of the draft implementation: https://gist.github.com/Chootin/2a73bdad96c2acbf4172cafe1d8745f8

Any thoughts?

Reply:
Started to play around with this, but haven't finished. Did do some research about the possibility of the delegate object being moved as well ([see the other comment](https://github.com/ros2-dotnet/ros2_dotnet/pull/124#discussion_r1385619713)), as the implementation needs to consider this as well.

Either we need to pin the delegate object that is passed to `rcl_clock_add_jump_callback` using a `GCHandle` or we use one static delegate (pinned as well?) and pass some integer id as `user_data` to do the lookup to the `JumpHandler`. Will look at those some more the next days.

> Also is the intention for users to declare a class variable to keep the IDisposable in scope within their own code or are we going to hold onto it in a list on Clock to prevent the Garbage Collector from automatically disposing of it?

I wouldn't remove the callback once the IDisposable gets collected by the CG as this causes indeterministic behaviour. If the IDisposable doesn't get saved the caller has no change other then disposing the clock to remove the callback. This is how `rclpy` behaves as far as I have seen it. `rclcpp` usses a SharedPointer, but there the Destructor gets called deterministically. So I would stick with how `rclpy` does it in this case.

Reply:
I now think I have a solution that fits thread safety and locking, see this commit here: https://github.com/hoffmann-stefan/ros2_dotnet/commit/75f6cf2753e93d597c70896102890d02c04e2763 ([branch](https://github.com/hoffmann-stefan/ros2_dotnet/commits/feature/clock-timer))

This is based on your draft, here are some notes what and why I changed things:

- Removed the Finalizer on `JumpHandler`.
  - We don't want indeterministic behavoir.
  - It would be really hard to implement right.
- Therefore the `_registered` flag is not needed to avoid `rcl_clock_remove_jump_callback` being called by the Finalizer.
- `private JumpCallbackInternal JumpCallback => OnJump;`: created a new Delegate each time. Now only one Delegate is created in the constructor explicitly. This is needed so PInvoke uses the same function pointer stub every time.
- `_registeredJumpCallbacks` is used to keep a strong reference so that the function pointer stub doesn't get deallocated.
- `_dispose` flag is set inside the lock, same for modifying `_registeredJumpCallbacks`
- my advice on removing the book-keeping `_registededJumpCallbacks` was wrong given the discussion in https://github.com/ros2-dotnet/ros2_dotnet/pull/124/files#r1379212544. We need to keep a strong reference to the delegate. We can't trust the caller of the `CreateJumpCallback` to do so.
- `_registededJumpCallbacks` was changed to a `HashSet`
- added a try/catch block so that exceptions from the callback don't propagate to native functions.
- Renamed `AddJumpCallback` to `CreateJumpCallback` to be consistent with `rclpy`/`rclcpp`.
- Did try to document the way things are done explicitly.

Didn't expect this change to be that complicated...

Would you check this and cherry-pick that commit if it looks good to you?